django_mongodb/compiler.py

```diff
@@ -0,0 +1,291 @@
+from django.conf import settings
+from django.core.exceptions import EmptyResultSet
+from django.db import (
+    DatabaseError,
+    IntegrityError,
+    NotSupportedError,
+    connections,
+)
+from django.db.models import NOT_PROVIDED, Count, Expression
+from django.db.models.constants import LOOKUP_SEP
+from django.db.models.sql import compiler
+from django.db.models.sql.constants import GET_ITERATOR_CHUNK_SIZE, MULTI
+
+from .base import Cursor
+from .query import MongoQuery, safe_call
+
+
+class SQLCompiler(compiler.SQLCompiler):
+    """Base class for all Mongo compilers."""
+
+    query_class = MongoQuery
+
+    def execute_sql(
+        self, result_type=MULTI, chunked_fetch=False, chunk_size=GET_ITERATOR_CHUNK_SIZE
+    ):
+        self.pre_sql_setup()
+        # QuerySet.count()
+        if self.query.annotations == {"__count": Count("*")}:
+            return [self.get_count()]
+        try:
+            query = self.build_query()
+        except EmptyResultSet:
+            return None
+        return query.fetch()
+
+    def results_iter(
+        self,
+        results=None,
+        tuple_expected=False,
+        chunked_fetch=False,
+        chunk_size=GET_ITERATOR_CHUNK_SIZE,
+    ):
+        """
+        Return an iterator over the results from executing query given
+        to this compiler. Called by QuerySet methods.
+        """
+        columns = self.get_columns()
+
+        if results is None:
+            # QuerySet.values() or values_list()
+            try:
+                results = self.build_query(columns).fetch()
+            except EmptyResultSet:
+                results = []
+
+        for entity in results:
+            yield self._make_result(entity, columns, tuple_expected=tuple_expected)
+
+    def has_results(self):
+        return bool(self.get_count(check_exists=True))
+
+    def _make_result(self, entity, columns, tuple_expected=False):
+        """
+        Decode values for the given fields from the database entity.
+
+        The entity is assumed to be a dict using field database column
+        names as keys.
+        """
+        result = []
+        for col in columns:
+            field = col.field
+            column = col.target.column
+            value = entity.get(column, NOT_PROVIDED)
+            if value is NOT_PROVIDED:
+                value = field.get_default()
+            else:
+                # Decode values using Django's database converters API.
+                converters = self.connection.ops.get_db_converters(col) + field.get_db_converters(
+                    self.connection
+                )
+                for converter in converters:
+                    value = converter(value, col, self.connection)
+            result.append(value)
```

Comment:
It was on my mind to optimize this a bit more. Take a look at commit "fetch database converters once per query rather than for each result." To answer your question, there could be both types of converters. The logic comes from `django.db.models.sql.SQLCompiler.get_converters()`.